src/@astro/internal/index.ts

```diff
@@ -18,11 +18,24 @@ export async function renderToString(result: any, componentFactory: AstroCompone
 
 export async function renderPage(result: any, Component: AstroComponentFactory, props: any, children: any) {
   const template = await renderToString(result, Component, props, children);
-  const styles = Array.from(result.styles).map(style => `<style>${style}</style>`);
-  const scripts = Array.from(result.scripts);
+  const styles = Array.from(result.styles).map(style => renderElement('style', style as any));
+  const scripts = Array.from(result.scripts).map(script => renderElement('script', script as any));
   return template.replace("</head>", styles.join('\n') + scripts.join('\n') + "</head>");
 }
 
+function renderElement(name: string, { props: _props, children = ''}: { props: Record<any, any>, children?: string }) {
+  const { hoist: _, "data-astro-id": astroId, "define:vars": defineVars, ...props } = _props;
+  if (defineVars) {
+    if (name === 'style') {
+      children = defineStyleVars(astroId, defineVars) + '\n' + children;
+    }
+    if (name === 'script') {
+      children = defineScriptVars(defineVars) + '\n' + children;
+    }
+  }
+  return `<${name}${spreadAttributes(props)}>${children}</${name}>`
+}
+
```

Comment:
Pretty straightforward function that renders `{props:{}, children: ""}` to an HTML string.